### DISCUSSION TITLE
Add a special 'timer' step option to recipes

### DISCUSSION THREAD

**Discussion by tommyalatalo** [mealie-recipes/mealie#1573] [Feature Request]
**Before submitting this feature request I have**
_COMPLETE AND FILL THESE OUT_
- [X] Reviewed the documentation and verified that my feature request isn't already a feature.
- [X] Reviewed the existing feature requests and my feature has not been requested.
- [X] Checked the [tasks tagged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
When following a recipe it would be very useful to have timers as part of the cooking tasks so that you can not only follow the manual steps of preparing ingredients etc. but also manage the time for cooking using the same list of steps.

**(Optional): Suggest A Solution**
Implement a special `timer` type step to the list of cooking tasks.
The timer step could contain a `Start Timer` button, with a predefined time that you set when editing the recipe.

Since adjusting the time on the fly is quite useful it might be good to

- A: Optionally have a `+ -` button in the step field which allows you to increment the time on the timer.
- B: Have the "Start Timer" button open up a dialog which allows the user to adjust the time before starting it.

When the timer ends it would be great to have it sound an alarm.

Example of cooking steps:

```
Step 1: Add ingredient X and Y, mix them in a bowl.
Step 2: Pour them into a pan and boil for 10 minutes.
Step 3: <button>Start Timer</button> (preset to 10 minutes)
Step 4: Turn stove off and let cool
```

**Additional Information**
- [ ] If the feature is accepted I'm willing to submit a PR to provide this feature

**Comment by epheterson** [comment]
This is a cool idea, it could even read the text of the steps and suggest a time.

**Comment by KidA90** [comment]
I recently migrated from Paprika Recipe Manager to Mealie and although Mealie does a better job for most of my requirements, this is a feature I really really miss compared to Paprika! I was quite happy with the functions in Paprika, so I want to give you a quick overview of them:
- auto-detect time measurements in the recipe
- start a respective timer by clicking on the time measurement
- manage running timers system-wide and manually add/delete timers

Especially the last function comes in pretty handy if you cook multiple recipes at once (think of the upcoming holidays ;-) or simply multitask within one recipe.

I attached some screenshots from Paprika to give you an idea what the function looks like. It would be great to have something similar in Mealie!
![1_small](https://user-images.githubusercontent.com/58784953/200197244-fb14eec9-1500-4ec5-b9ed-42749676a0fe.png)
![2_small](https://user-images.githubusercontent.com/58784953/200197247-7a406f4b-90c5-4464-a08a-177b2c551630.png)
![3_small](https://user-images.githubusercontent.com/58784953/200197248-22018984-56d7-4d37-baf8-b25b0b83ed87.png)
![4_small](https://user-images.githubusercontent.com/58784953/200197249-6bf53b6d-26ee-4357-8b9e-724c67e33550.png)

**Reply by Inervo** [reply]
Coming from Paprika as well, this feature is really a must, and I would love to have it on Mealie :)

**Reply by felixschndr** [reply]
This looks really useful

**Comment by josjojo** [comment]
I also like this feature. Maybe it can parse time in the text to change it to a hyperlink. If you parse text it needs to take the language into account.

**Comment by atiris** [comment]
In my opinion, it would be convenient to **connect the timer to a specific step** (record in recipeInstructions), but not to create another step "category". Some steps do not need a timer (for example: mix all the liquid ingredients until combined), others have a fixed timer (for example: beat for 5 minutes) and others require a timer in several steps (for example: first bake for 10 minutes at 230 degrees, then ventilate and bake for another 20 minutes at 180 degrees). The timer should not be a separate step in the recipe, because any such timer would belong to some step before it - logically, it would be a wrong choice in the structure of the application. Rather, I think that the **duration is an attribute of the given step**.

So something like this:

    {
      "id": "...",
      "title": "...",
      "text": "...",
      "duration": "00:21:00",
      "timers": ["00:15:00", "00:05:00"],
      "ingredientReferences": [...]
    },

With such a structure, the application could also expect an audio instruction, for example, "Mealie, next" after a certain amount of time in **duration** for a running step is done, or be reminded (by a sound, tick or something like this) that another step should follow and wait for user activity.

A **timers** item would allow starting a basic timer and continuing with another if the given step has more actions. But even adding an attribute for one timer per step would benefit the recipe structure. When importing the recipe and parsing, it is possible to extract the timers from the text.

Currently, only *"Step: X"* is displayed in the title bar of each step. With the option to write step durations and add timers, it could look like this: *"Step: X [Start Step icon] [Start Timer icon]"* Clicking on title or Start step button can starts a countdown of the step time, something like a Progress indicator above the given step. Clicking Timer starts first timer. Clicking elsewhere would end the step and mark it as done, as it is now.

Regarding the design and functionality of the current timer, I see the **following opportunities** for improvement:
- the **design is not connected** to the rest of the app (it looks like it was designed by someone other than the app)
- **timer is not linked** to the steps in the recipe
- it is **not visible when scrolling** through the recipe (this is a big design flaw, if I want to start the timer in a later step I have to scroll all the way to the top to start and go back...) the timer should always be available as a floating button
- it is** not possible to turn it off if it is minimized** without scrolling up (similar to the previous step, and I watch the step below again and the timer starts, I have to scroll up and press the button)
- the timer **does not display information about the remaining time** without clicking it (if we already have a timer icon somewhere, it might be possible to write inside, for example: 10m or 3m or 45s according to the remaining time).
- does **not allow to quickly add a default time** (plus, minus a minute, plus minus 5 minutes, multiple presses will add more time)

Sorry, I'm not a designer (definitely don't think of this as a solution suggestion), I just rough sketched in the demo app what I mean from a timer point of view in a given step and what a floating countdown button would look like.

*Timer for 45 minutes remaining*
<img src="https://github.com/user-attachments/assets/762de00f-4148-4c72-b181-e7a9972295c8" width="360">

*Timer for 28 seconds left*
<img src="https://github.com/user-attachments/assets/b123171f-7db9-4eff-8c78-e95a747e0f80" width="360">

*Scrolled position of timer with 2 hours remaining and 2 predefined step timers*
<img src="https://github.com/user-attachments/assets/3640c913-a3f3-4849-af6b-6f8565eb9bd3" width="360">

**Reply by Inervo** [reply]
Love your proposition mate!!
I hope this will be in the next releases <3 That's the BIG thing laking currently

**Reply by CloCkWeRX** [reply]
I recently added a PR for reading the steps (#4997), adding the "next" verbal command would be a logical progression of this.

**Comment by codetakki** [comment]
I have made a pull request with my suggestion, it adds a  timer directly to the instructions, based on the text  if 5 min etc are detected: #4467
